Migrating GenerateContentIdService

Showtim3 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [X] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `python -m scripts.run_frontend_tests`.
- [X] The PR is made from a branch that's **not** called "develop".
- [X] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [X] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [ ] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [ ] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [ ] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

Showtim3 committed:
migrating GenerateContentIdService

Showtim3 commented on core/templates/dev/head/services/GenerateContentIdServiceSpec.ts:
```diff
@@ -60,6 +44,7 @@ describe('GenerateContentIdService', function() {
 
   it('should throw error for unknown content id', function() {
     expect(function() {
+      // @ts-ignore
```
This is intentional since we are explicitly passing a single parameter whereas the functions expects two paraments. We want that function to throw an error.

DubeySandeep replied:
You can add this description to top of this line.

Also, do you know why we are passing one argument here, I think passing two-argument will also work, right?

DubeySandeep replied:
I don't think passing one parameter over here is intentional, as we are testing the error not the absence of parameter. Can you please add all the parameters and remove the `@ts-ignore`?

DubeySandeep commented on core/templates/dev/head/services/GenerateContentIdService.ts:
```diff
@@ -17,40 +17,48 @@
-          var tempCount =
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+import { AppConstants } from '../app.constants';
```
Change this to:
`import { AppConstants } from 'app.constants';
`
Also add a newline above.

Showtim3 replied:
Done

DubeySandeep commented on core/templates/dev/head/services/GenerateContentIdService.ts:
```diff
@@ -17,40 +17,48 @@
+  providedIn: 'root'
+})
+export class GenerateContentIdService {
+  generateIdForComponent(existingComponentIds: any,
```
Why `existingComponentIds` is of type `any`?

Showtim3 replied:
Added a todo! Done

DubeySandeep replied:
@Showtim3 I think it would be quiet easy to find the type of these value, can you please check?

Also, do let me know if you are finding it hard to find the type of these parameters, I can help you! :)

DubeySandeep commented on core/templates/dev/head/services/GenerateContentIdService.ts:
```diff
@@ -17,40 +17,48 @@
+})
+export class GenerateContentIdService {
+  generateIdForComponent(existingComponentIds: any,
+      componentName: string): any {
```
Why the return type is any?

Showtim3 replied:
Added a todo. Done

DubeySandeep commented on core/templates/dev/head/services/GenerateContentIdService.ts:
```diff
@@ -17,40 +17,48 @@
+    return (searchKey + String(count + 1));
+  }
+
+  _getNextId(existingComponentIds: any, componentName: string): any {
```
Ditto why the type is `any`? 

In general, try adding a comment on top wherever you use `any` type for arguments.

Showtim3 replied:
Done

DubeySandeep commented on core/templates/dev/head/services/GenerateContentIdService.ts:
```diff
@@ -17,40 +17,48 @@
+    }
+  }
+
+  getNextId(existingComponentIds: any, componentName: string): any {
```
Ditto*

Showtim3 replied:
Done

DubeySandeep reviewed (changes requested):
@Showtim3, Thanks for the PR!  I took a pass and left a few comments PTAL!

In general, if you think some line of code needs explanation then try adding a comment in the file instead of just explaining it to the reviewer on GitHub as you are expecting that code to be merged and adding a comment in the file help new readers to understand the code.

DubeySandeep commented:
@Showtim3, Can you please update the PR title and explanation as asked in the `Checklist`?

kevinlee12 commented on core/templates/dev/head/services/GenerateContentIdServiceSpec.ts:
```diff
@@ -60,6 +44,7 @@ describe('GenerateContentIdService', function() {
 
   it('should throw error for unknown content id', function() {
```
Please change the frontend tests from `function() {` to `() => {`

Showtim3 replied:
Done

kevinlee12 reviewed (changes requested):
@Showtim3, along with the comments Sandeep had, I have another one too. ptal!

Showtim3 committed:
adding todo for any

kevinlee12 commented on core/templates/dev/head/services/GenerateContentIdService.ts:
```diff
@@ -17,40 +17,48 @@
+    }
+    return (searchKey + String(count + 1));
+  }
+  // TODO(#7176): Replace 'any' with the exact type.
```
Add newline above this comment.

Showtim3 replied:
Done

kevinlee12 commented on core/templates/dev/head/services/GenerateContentIdService.ts:
```diff
@@ -17,40 +17,48 @@
+      throw Error('Unknown component name provided.');
+    }
+  }
+  // TODO(#7176): Replace 'any' with the exact type.
```
Add newline above this comment.

Showtim3 replied:
Done

kevinlee12 reviewed (approved):
couple of nits; otherwise lgtm!

kevinlee12 commented:
@DubeySandeep code owner review please!

Showtim3 committed:
adding new lines

Showtim3 committed:
fixing any data types

Showtim3 committed:
merging develop

Showtim3 commented:
@DubeySandeep Thanks for ur feedback, I have added the return types for all, Also I have added a comment in the code regarding the ts-ignore thing. And yeah we can also pass two parameters for generating an error, but the test might have been written with the intention that we want to test if the function is being called with only a single parameter then an error should be thrown. If you suggest, i can change the test to pass 2 parameters.

kevinlee12 commented on core/templates/dev/head/services/GenerateContentIdService.ts:
```diff
@@ -17,40 +17,48 @@
+  providedIn: 'root'
+})
+export class GenerateContentIdService {
+
```
Remove empty line

Showtim3 replied:
Done

kevinlee12 commented:
@Showtim3 mind the failing lint tests!

Showtim3 committed:
removing empty line

DubeySandeep commented on core/templates/dev/head/services/GenerateContentIdService.ts:
```diff
@@ -17,40 +17,47 @@
+  providedIn: 'root'
+})
+export class GenerateContentIdService {
+  generateIdForComponent(existingComponentIds: Object,
```
`existingComponentIds` should be array of strings right? (Check the test file)

Showtim3 replied:
Done

DubeySandeep commented on core/templates/dev/head/services/GenerateContentIdService.ts:
```diff
@@ -17,40 +17,47 @@
+      componentName: string): string {
+    let contentIdList = JSON.parse(JSON.stringify(existingComponentIds));
+    let searchKey = componentName + '_';
+    let count = 0;
```
Do we need to define the type of the local (block scope) variable as well?

/cc @bansalnitish

bansalnitish replied:
Yes, we should do this ideally. Tbh, there are so many variables, function arguments, function return types that adding types to each thing will definitely effect the pace. I think we need to discuss this once (Also, this should've been done during typescript migration ideally but that's fine).

What do you think @kevinlee12 @seanlip @Showtim3 @ankita240796?

seanlip replied:
I strongly think we should do typing for variables and return types in public APIs, but I feel that typing for local variables can happen in a separate, later phase. That said, I would appreciate understanding what others' thoughts are on this, too. 

@kevinlee12 @Showtim3 @ankita240796 what do you think?

kevinlee12 replied:
Agreed, we can type variables inside functions later, but we should type public APIs (args and returns) now.

Showtim3 replied:
Agreed! And yeah we are adding the args and the return type's. So we got it covered.
and adding types for local variables would be really nice to have, but as per our goals this doesn't feasible at this point of time.

bansalnitish replied:
Sounds good to me too!

DubeySandeep commented on core/templates/dev/head/services/GenerateContentIdService.ts:
```diff
@@ -17,40 +17,47 @@
-      }
-    };
-  }]);
+  _getNextId(existingComponentIds: Object, componentName: string): string {
```
(Ditto) can you please check the type of `existingComponentIds`?

Showtim3 committed:
adding Array of strings, removing ts-ignore

Showtim3 committed:
fixed imports

Showtim3 committed:
merging develop

DubeySandeep reviewed (approved):
LGTM is a code owner!

@bansalnitish, PTAL!

Showtim3 committed:
merging develop

Showtim3 commented:
Closing in reference to #7934
